pkg/manager/impl/reservation_manager.go

```diff
@@ -0,0 +1,189 @@
+	getTagFailure                labeled.Counter
+}
+
+type NowFunc func() time.Time
+
+type reservationManager struct {
+	repo               repositories.RepositoryInterface
+	reservationTimeout time.Duration
+	now                NowFunc
+	systemMetrics      reservationMetrics
+}
+
+func NewReservationManager(
+	repo repositories.RepositoryInterface,
+	reservationTimeout time.Duration,
+	nowFunc NowFunc, // Easier to mock time.Time for testing
+	reservationScope promutils.Scope,
+) interfaces.ReservationManager {
+	systemMetrics := reservationMetrics{
+		scope: reservationScope,
+		reservationAcquired: labeled.NewCounter(
+			"reservation_acquired",
+			"Number of times a reservation was acquired",
+			reservationScope),
+		reservationAlreadyInProgress: labeled.NewCounter(
+			"reservation_already_in_progress",
+			"Number of times we try of acquire a reservation but the reservation is in progress",
+			reservationScope,
+		),
+		acquireReservationFailure: labeled.NewCounter(
+			"acquire_reservation_failure",
+			"Number of times we failed to acquire reservation",
+			reservationScope,
+		),
+		getTagFailure: labeled.NewCounter(
+			"get_tag_failure",
+			"Number of times we failed to get tag",
+			reservationScope,
+		),
+	}
+
+	return &reservationManager{
+		repo:               repo,
+		reservationTimeout: reservationTimeout,
+		now:                nowFunc,
+		systemMetrics:      systemMetrics,
+	}
+}
+
+func (r *reservationManager) GetOrReserveArtifact(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (*datacatalog.GetOrReserveArtifactResponse, error) {
+	tagKey := transformers.ToTagKey(request.DatasetId, request.TagName)
+	tag, err := r.repo.TagRepo().Get(ctx, tagKey)
+	if err != nil {
+		if errors2.IsDoesNotExistError(err) {
+			// Tag does not exist yet, let's acquire the reservation to work on
+			// generating the artifact.
+			status, err := r.tryAcquireReservation(ctx, request)
+			if err != nil {
+				r.systemMetrics.acquireReservationFailure.Inc(ctx)
+				return nil, err
+			}
+
+			return &datacatalog.GetOrReserveArtifactResponse{
+				Value: &datacatalog.GetOrReserveArtifactResponse_ReservationStatus{
+					ReservationStatus: &status,
+				},
+			}, nil
+		}
+		logger.Errorf(ctx, "Failed retrieve tag: %+v, err: %v", tagKey, err)
+		r.systemMetrics.getTagFailure.Inc(ctx)
+		return nil, err
+	}
+
+	artifact, err := transformers.FromArtifactModel(tag.Artifact)
+	if err != nil {
+		return nil, err
+	}
+
+	return &datacatalog.GetOrReserveArtifactResponse{
+		Value: &datacatalog.GetOrReserveArtifactResponse_Artifact{
+			Artifact: artifact,
+		},
+	}, nil
+}
+
+func (r *reservationManager) tryAcquireReservation(ctx context.Context, request *datacatalog.GetOrReserveArtifactRequest) (datacatalog.ReservationStatus, error) {
+	repo := r.repo.ReservationRepo()
+	reservationKey := transformers.ToReservationKey(*request.DatasetId, request.TagName)
+	rsv, err := repo.Get(ctx, reservationKey)
```

Comment:
Why do we try to do a `Get` here? Why do we not always try to CreateOrUpdate? is it to optimize the write?

Reply:
Yeah this is an optimization to reduce the number of writes to db. We always need to do a `GET` here because we want to know who owns the reservation and show it to users on the UI. However, the reservation is held by a single task most of the times and there is no need to do a write.

Reply:
awesome, please add some docs here to explain that..

Reply:
done